### PR TITLE
Add fault tolerance for the StreamingDataset 1/n

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds support for fault tolerance for the StreamingDataset. This enables to use it seamlessly with Fabric.save as follows:

```python
dataset = StreamingDataset(...)

state = {...., "model": model, "dataset": dataset}

...

fabric.save(state)
```

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
:books: Documentation preview :books:: https://pytorch-lightning--19049.org.readthedocs.build/en/19049/

### CODE REVIEW

File: tests/tests_data/streaming/test_dataset.py [hunk 1]
```diff
@@ -527,3 +530,199 @@ def test_s3_streaming_dataset():
     dataset = StreamingDataset(input_dir="s3://pl-flash-data/optimized_tiny_imagenet")
     assert dataset.input_dir.url == "s3://pl-flash-data/optimized_tiny_imagenet"
     assert dataset.input_dir.path is None
+
+
+@pytest.mark.skipif(sys.platform == "win32", reason="Not tested on windows and MacOs")
+def test_resumable_dataset_single_worker(tmpdir):
```
Comment (lantiga) [inline tests/tests_data/streaming/test_dataset.py:536]: I saw we are not testing the effect of `current_epoch` when resuming, maybe worth adding

**Review by lantiga — APPROVED** [review]
Looks great! A couple of comments

File: src/lightning/data/streaming/dataset.py [hunk 2]
```diff
@@ -188,15 +232,159 @@ def __next__(self) -> Any:
...
+                tmp,
+            )
+
+            # 2. Flush to make sure it is written
+            tmp.flush()
+
+            # 3. Move the file to avoid corrupted read from the main thread.
+            now = datetime.now().strftime(_TIME_FORMAT)
+            checkpoint_path = os.path.join(self.cache.checkpoint_rank_dir, f"checkpoint-{now}.json")
+            shutil.copyfile(tmp.name, checkpoint_path)
```
[the remark this answers is not in the document]
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:286]: Good catch !

File: tests/tests_data/streaming/test_dataset.py (hunk 1 above)
Comment (tchaton) [reply tests/tests_data/streaming/test_dataset.py:536]: Done !

File: src/lightning/data/streaming/cache.py [hunk 3]
```diff
@@ -94,6 +94,10 @@ def __init__(
         self._is_done = False
         self._distributed_env = _DistributedEnv.detect()
 
+    @property
+    def rank(self) -> int:
+        return self._reader.rank
```
Comment (awaelchli) [inline src/lightning/data/streaming/cache.py:99]: Why would this make sense? The rank is something global, accessing it through a particular object doesn't make much sense. We should always use the distributed env for this. It's the source of truth.

File: src/lightning/data/streaming/cache.py [hunk 4]
```diff
@@ -102,6 +106,20 @@ def filled(self) -> bool:
         self._is_done = os.path.exists(os.path.join(self._cache_dir, _INDEX_FILENAME))
         return self._is_done
 
+    @property
+    def checkpoint_dir(self) -> str:
```
Comment (awaelchli) [inline src/lightning/data/streaming/cache.py:110]: It shouldn't be necessary to duplicate the code in both of these.

File: src/lightning/data/streaming/dataset.py [hunk 5]
```diff
@@ -42,6 +56,7 @@ def __init__(
         drop_last: bool = False,
         seed: int = 42,
         serializers: Optional[Dict[str, Serializer]] = None,
+        checkpoint_interval: int = 60 * 5,
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:59]: The caveat here is that 1) the checkpoint might be outdated when we call state_dict() if the interval is too large and 2) the workers may have saved progress for a sample that has not yet been collated into a batch and fetched by the user.

**Review by Borda — APPROVED** [review]

File: src/lightning/data/streaming/dataset.py [hunk 6]
```diff
@@ -141,10 +170,26 @@ def __iter__(self) -> "StreamingDataset":
...
+            state = self._state_dict[str(self.cache.rank)]
+
+            # re-generate indexes
+            interval = self.worker_intervals[self.chunk_index]
+            current_indexes = np.arange(interval[0], interval[1])
+            current_indexes = self.shuffler(current_indexes, self.current_epoch, self.chunk_index)
+            self.current_indexes = current_indexes[state["index"] :]
+
+            # Bump the chunk_index
+            self.chunk_index += 1
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:184]: Why +1 the index? We're reloading it in the line above. If the chunk wasn't complete, we would now miss the remainder?

File: src/lightning/data/streaming/dataset.py [hunk 7]
```diff
@@ -188,15 +238,165 @@ def __next__(self) -> Any:
...
+                        "world_size": self.distributed_env.world_size,
+                        "num_workers": self.worker_env.world_size,
+                        "shuffle": self.shuffle,
+                    },
+                    f,
+                )
+
+            # 3. Move the file to avoid corrupted read from the main thread.
+            now = datetime.now().strftime(_TIME_FORMAT)
+            checkpoint_path = os.path.join(self.cache.checkpoint_rank_dir, f"checkpoint-{now}.json")
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:289]: why is the file time stamped? This will create so many files. IMO we should overwrite the file, because only the current state matters. And since every worker only saves to their dedicated folder, it is safe.

File: src/lightning/data/streaming/dataset.py [hunk 8]
```diff
@@ -188,15 +238,165 @@ def __next__(self) -> Any:
...
+        # Checkpoint based on time
+        if (self.last_time - time()) > self.checkpoint_interval:
+            self.checkpoint(self.chunk_index - 1)
+
         return data
 
+    def checkpoint(self, chunk_index: int) -> None:
+        # Checkpointing isn't supported for windows
+        if sys.platform == "win32":
+            return
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:258]: Then we should raise an error.

File: src/lightning/data/streaming/dataset.py [hunk 9]
```diff
@@ -188,15 +238,165 @@ def __next__(self) -> Any:
                 chunk_index=self.worker_chunks[self.chunk_index - 1],
                 # We provide the chunks indexes only one the first
                 chunk_indexes=None if self.has_triggered_download else self.worker_chunks,
-                last_index=last_index,
+                last_index=(self.chunk_index - 1) == len(self.worker_intervals) and len(self.current_indexes) == 1,
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:241]: consider renaming this to `is_last_index`, because the name suggests it is an int.

File: src/lightning/data/streaming/dataset.py [hunk 10]
```diff
@@ -188,15 +238,165 @@ def __next__(self) -> Any:
...
+        self.global_index += 1
         self.index += 1
 
+        # Checkpoint based on time
+        if (self.last_time - time()) > self.checkpoint_interval:
+            self.checkpoint(self.chunk_index - 1)
+
         return data
 
+    def checkpoint(self, chunk_index: int) -> None:
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:255]: This shouldn't be public. The user wouldn't be able to use this effectively, because they can only call it from the main process. And from the main process it never makes sense. 

I suggest to  1) make it private 2) raise an error if called in main process

File: src/lightning/data/streaming/dataset.py [hunk 11]
```diff
@@ -188,15 +238,165 @@ def __next__(self) -> Any:
...
+                    if node_rank in node_ranks:
+                        continue
+                    state = {}
+                    obj = [_state_dict]
+                    torch.distributed.broadcast_object_list(obj, index, group=_group.WORLD)
+                    state = obj[0]
+                    state_dict.update(**state)
+                    node_ranks.append(node_rank)
+        else:
+            raise NotImplementedError("The `state_dict` should be called on the main thread.")
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:335]: But they aren't threads, they are proper processes. And we should raise immediately at the beginning, this would eliminate the entire if-else block, making the code much more readable

File: src/lightning/data/streaming/dataset.py [hunk 12]
```diff
@@ -188,15 +238,165 @@ def __next__(self) -> Any:
...
+                for index in range(self.distributed_env.world_size):
+                    node_rank = index // num_devices
+                    if node_rank in node_ranks:
+                        continue
+                    state = {}
+                    obj = [_state_dict]
+                    torch.distributed.broadcast_object_list(obj, index, group=_group.WORLD)
+                    state = obj[0]
+                    state_dict.update(**state)
+                    node_ranks.append(node_rank)
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:333]: we should put this in a function. Way easier to unit test!

File: src/lightning/data/streaming/dataset.py [hunk 13]
```diff
@@ -188,15 +238,165 @@ def __next__(self) -> Any:
...
+            self.cache = self._create_cache(worker_env=self.worker_env)
+
+        state_dict: Dict[str, Any] = {}
+        worker_env = _WorkerEnv.detect()
+        if worker_env.world_size == 1:
+            # 1. Check whether the checkpoint_dir exists
+            if not os.path.exists(self.cache.checkpoint_dir):
+                return state_dict
+
+            # 2. Iterate through the workers and read the latest checkpoint
```
Comment (awaelchli) [inline src/lightning/data/streaming/dataset.py:308]: This step wouldn't be necessary, see comment above

File: src/lightning/data/streaming/shuffle.py [hunk 14]
```diff
@@ -147,6 +144,5 @@ def get_chunks_and_intervals_per_ranks(self, distributed_env: _DistributedEnv, c
 
         return chunks_per_ranks, intervals_per_ranks
 
-    def __call__(self, array: np.ndarray) -> List[int]:
-        assert self.random_state
-        return self.random_state.permutation(array).tolist()
+    def __call__(self, array: np.ndarray, current_epoch: int, chunk_index: int) -> List[int]:
+        return np.random.RandomState(seed=self.seed + current_epoch + chunk_index).permutation(array).tolist()
```
Comment (awaelchli) [inline src/lightning/data/streaming/shuffle.py:148]: This is problematic, because the seed will not be unique. For example 5 + 4 = 4 + 5 = 9. We should at least multiple the current epoch by number of chunks.

File: src/lightning/data/streaming/cache.py (hunk 3 above)
Comment (tchaton) [reply src/lightning/data/streaming/cache.py:99]: Done !

File: src/lightning/data/streaming/cache.py (hunk 4 above)
Comment (tchaton) [reply src/lightning/data/streaming/cache.py:110]: Sure.

File: src/lightning/data/streaming/dataset.py (hunk 5 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:59]: Yes. Without patching the DataLoader, this is the best proxy. It won't be perfect but gives a good approximation for a V0.
We can revisit later on.

Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:59]: Putting it down to 1 min.

File: src/lightning/data/streaming/dataset.py (hunk 6 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:184]:
```python
  state = self._state_dict[str(self.cache.rank)]

  # re-generate indexes
  interval = self.worker_intervals[self.chunk_index]
  current_indexes = np.arange(interval[0], interval[1])
  current_indexes = self.shuffler(current_indexes, self.current_epoch, self.chunk_index)
  self.current_indexes = current_indexes[state["index"] :]

  # Bump the chunk_index
  self.chunk_index += 1
```

No, it won't. The `chunk_index` is bumped only once the  `current_indexes` are re-created.

I will clean this up in another PR.

File: src/lightning/data/streaming/dataset.py (hunk 8 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:258]: Done !

File: src/lightning/data/streaming/dataset.py (hunk 9 above)
Comment (tchaton) [reply src/lightning/data/streaming/dataset.py:241]: Done !
